src/core/DataFile.cpp

```diff
@@ -328,6 +405,164 @@ bool DataFile::writeFile( const QString& filename )
+					{
+						resPath = QFileInfo(m_fileName).path() + "/" + resPath.remove(0,
+							PathUtil::basePrefix(PathUtil::Base::LocalDir).length());
+					}
+
+					// Check if we need to add a counter to the filename
+					QString finalFileName = QFileInfo(resPath).fileName();
+					QString extension = resPath.section('.', -1);
+					int repeatedNames = 0;
+					for (QString name : namesList)
+					{
+						if (finalFileName == name)
+						{
+							++repeatedNames;
+						}
+					}
+					// Add the name to the list before modifying it
+					namesList.push_back(finalFileName);
+					if (repeatedNames)
+					{
+						// Remove the extension, add the counter and add the
+						// extension again to get the final file name
+						finalFileName.truncate(finalFileName.lastIndexOf('.'));
+						finalFileName = finalFileName + "-" + QString::number(repeatedNames) + "." + extension;
+					}
+
+					// Final path is our resources dir + the new file name
+					QString finalPath = resourcesDir + "/" + finalFileName;
+
+					// Copy resource file to the resources folder
+					if(!QFile::copy(resPath, finalPath))
+					{
+						qWarning("ERROR: Failed to copy resource");
+						return false;
+					}
+
+					// Update attribute path to point to the bundle file
+					QString newAtt = PathUtil::basePrefix(PathUtil::Base::LocalDir) + "resources/" + finalFileName;
+					el.setAttribute(*res, newAtt);
+				}
+				++res;
+			}
+		}
+		++it;
+	}
+
+	return true;
+}
+
+
+
+
+/* @brief This recursive method will go through all XML nodes of the DataFile
+ *        and check whether any of them have local paths. If they are not on
+ *        our list of elements that can have local paths we return true,
+ *        indicating that we potentially have plugins with local paths that
+ *        would be a security issue. The Song class can then abort loading
+ *        this project.
+ * @param QDomElement parent The parent node being interated. When called
+ *        without arguments, this will be an empty element that will be
+ *        ignored (since the second parameter will be true).
+ * @param bool firstCall Defaults to true, and indicates to this recursive
+ *        method whether this is the first call. If it is it will use the
+ *        root element as the parent.
+ */
+bool DataFile::hasLocalPlugins(QDomElement parent /* = documentElement()*/, bool firstCall /* = true*/) const
+{
+	// If this is the first iteration of the recursion we use the root element
+	if (firstCall) { parent = documentElement(); }
+
+	auto children = parent.childNodes();
+	for (int i = 0; i < children.size(); ++i)
+	{
+		QDomNode child = children.at(i);
+		QDomElement childElement = child.toElement();
+		assert(childElement);
+
+		bool skipNode = false;
+		// Skip the nodes allowed to have "local:" attributes, but
+		// still check its children
+		for
+		(
+			ResourcesMap::const_iterator it = ELEMENTS_WITH_RESOURCES.begin();
```

Comment:
The idea of having a context dependent check is interesting, in case we turn out using those same tags in other contexts for different purposes. As you said it's quite a big change though, would probably have to rethink the `ELEMENT_WITH_RESOURCES` variable. It would have to contain not only the tag and allowed attributes, but also the parent tag. A good idea that we could keep track of to try implementing in the future, for now it's not strictly necessary but it could make the feature even more fail proof in case future changes allow this kind of thing.

I think the allowed tags are not top level though, aren't they children of tags like `<instrument>` and `<sampletrack>`?

Reply:
Oh correct, I got that wrong, they are not toplevel. OK, let's resolve this.